doc/developers/minimal_reproducer.rst

```diff
@@ -0,0 +1,409 @@
+MCVE from it.
+
+**Example**
+
+.. code-block:: python
+
+    # I am currently working in a ML project and when I tried to fit a
+    # GradientBoostingRegressor instance to my_data.csv I get a UserWarning:
+    # "X has feature names, but DecisionTreeRegressor was fitted without feature
+    # names". You can get a copy of my dataset from
+    # https://example.com/my_data.csv and verify my features do have
+    # names. The problem seems to arise during fit when I pass an integer
+    # to the n_iter_no_change parameter.
+
+    df = pd.read_csv('my_data.csv')
+    X = df[["feature_name"]] # my features do have names
+    y = df["target"]
+
+    # We set random_state=42 for the train_test_split
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.33, random_state=42)
+
+
+    # An instance with default n_iter_no_change raises no error nor warnings
+    gbdt = GradientBoostingRegressor(random_state=0)
+    gbdt.fit(X_train, y_train)
+
+    # the bug appears when I change the value for n_iter_no_change
+    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
+    gbdt.fit(X_train, y_train)
+
+Provide a failing code example with minimal comments
+----------------------------------------------------
+
+Instructions may get lost in translation for non-native english speakers.
+Hence, it is better that all the necessary information can be read
+from the code itself. Besides, by this point you already provided a concise
+description in the **Describe the bug** section of the Issue template.
+
+**Improved example**
+The following code, while **still not minimal**, is already **much better** because
+it can be copy-pasted in a python terminal to reproduce the problem in one step.
+In particular:
+- it contains **all necessary imports statements**;
+- it can fetch the public dataset without having to manually download a
+  file and put it in the expected location on the disk.
+.. code-block:: python
+
+    import pandas as pd
+    df = pd.read_csv("https://example.com/my_data.csv")
+    X = df[["feature_name"]]
+    y = df["target"]
+
+    from sklearn.model_selection import train_test_split
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.33, random_state=42)
+
+    from sklearn.ensemble import GradientBoostingRegressor
+
+    gbdt = GradientBoostingRegressor(random_state=0)
+    gbdt.fit(X_train, y_train)  # no warning
+
+    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
+    gbdt.fit(X_train, y_train) # raises warning
+
+
+Boil down your script to something as small as possible
+-------------------------------------------------------
+
+You have to ask yourself which lines of code are relevant and which are not for
+raising the bug. Erasing unnecessary lines of code will help you and other
+contributors narrow down the bug. In this example the warning has nothing to
+do with the `train_test_split`, nor the `random_state`.
```

Comment:
```suggestion
You have to ask yourself which lines of code are relevant and which are not for
reproducing the bug. Deleting unnecessary lines of code or simplifying the function
calls by omitting unrelated non-default options will help you and other contributors
narrow down the cause of the bug.

In particular, for  this specific example:

- the warning has nothing to do with the `train_test_split` since it already appears
   in the training set.
- similarly, the lines that compute the scores on the test set are not necessary;
- the bug can be reproduced for any value of`random_state` so leave it to its
   default;
- the bug can be reproduced without preprocessing the data with the
  `StandardScaler`.
```